design-proposals/shadow-node.md

```diff
@@ -0,0 +1,343 @@
+  | object.status == oldObject.status                                                                                                 | virt-handler user cannot modify status of the nodes                            |                                                                                                                                                                        |
+  | object.(not labels/annotations/resourceVersion/managedFields) == oldObject.(not labels/annotations/resourceVersion/managedFields) | virt-handler user can only change allowed sub-metadata fields.                 |                                                                                                                                                                        |
+  | size(`newNonKubevirtLabels`) == size(`oldNonKubevirtLabels`)                                                                      | virt-handler user cannot add/delete non kubevirt-owned labels                  |                                                                                                                                                                        |
+  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |
+
+### Backwards compatibility:
+validationAdmissionPolicy [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it should be OK for deployment on Kubevirt main and next minor release branches.
+However, since kubevirt needs to be supported up to 2 k8s versions back, then the deployment of these new objects needs to be conditioned by a new feature gate "node restriction". HCO main and v1.3 branches will enable this feature gate by default.
+
+### Next step: Prevent patches from other nodes' virt-handlers
+It is possible to expand the rules of the validationAdmissionPolicy to also prohibit virt-handler running on a node from patching other nodes.
+This could be defined by setting a new variable `requestHasNodeName` which will be true if request has `userInfo.extra.authentication.kubernetes.io/node-name` field, then adding this validation:
+```
+!`requestHasNodeName` OR (object.Name == request.nodeName)
+```
+However, this field is [beta](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#additional-metadata-in-pod-bound-tokens) on k8s 1.30  and is currently protected by the [ServiceAccountTokenPodNodeInfo](https://github.com/kubernetes/kubernetes/blob/c4bce63d9886e5f1fc00f8c3b5a13ea0d2bdf772/pkg/features/kube_features.go#L753) k8s feature gate.
```

Comment:
I guess this is considered a cons. Can you add it as such?

Reply:
Added.